# Lab | Intro to Probability

**Objective**

Welcome to this Intro to Probability lab, where we explore decision-making scenarios through the lens of probability and strategic analysis. In the business world, making informed decisions is crucial, especially when faced with uncertainties. This lab focuses on scenarios where probabilistic outcomes play a significant role in shaping strategies and outcomes. Students will engage in exercises that require assessing and choosing optimal paths based on data-driven insights. The goal is to enhance your skills by applying probability concepts to solve real-world problems.

**Challenge 1**

#### Ironhack Airlines 

Often Airlines sell more tickets than they have seats available, this is called overbooking. Consider the following:
- A plane has 450 seats. 
- Based on historical data we conclude that each individual passenger has a 3% chance of missing it's flight. 

If the Ironhack Airlines routinely sells 460 tickets, what is the chance that they have a seats for all passenger?

In [2]:
#raciocinio: 
# using scenario with ppl who MISS the flight

## p = 0.03   ->  3% each individual chance of missing the flight

## n= 460 (=460 passengers with tickets)

## k = how many persons missing 

## condition: at least 10 passengers have to miss the flight, since the plane has only 450 seats but tickets sold were 460
## 10 is my 'a' :
### rule: P(k >= a) -> P(k >= 10) 

## using cdf since k can have different values (so cdf = when accumulative)

## "10 or more" -> 1 - cdf(9)

## P(k >= a)  ->  1 - cdf(a - 1)
## so: P(k >= 10)  ->  1 - cdf(9)

In [3]:
from scipy.stats import binom

n = 460  
p = 0.03

result = 1 - binom.cdf(9, n, p)

print(f"Probability that everyone gets a seat: {result:.4f}")

Probability that everyone gets a seat: 0.8845


#### Based on the probability: 

##### There is an 88.45% chance that every passenger who shows up gets a seat.

##### This also means that in about 12% of the flights (100-88.45), more than 450 people will show up,so the airline will not have a seat for everyone.

##### Since on average only 3% of the passengers miss the flight, selling 460 tickets for 450 seats is a bit risky, but it works most of the times.

**Challenge 2**

#### Ironhack Call Center 

Suppose a customer service representative at a call center is handling customer complaints. Consider the following:
- The probability of successfully resolving a customer complaint on the first attempt is 0.3. 


What is the probability that the representative needs to make at least three attempts before successfully resolving a customer complaint?

In [3]:
from scipy.stats import geom

p = 0.3

result = 1 - geom.cdf(2, p)

print(f"Probability of needing at least 3 attempts: {result:.4f}")

Probability of needing at least 3 attempts: 0.4900


**Challenge 3**

#### Ironhack Website

Consider a scenario related to Ironhack website traffic. Where:
- our website takes on average 500 visits per hour.
- the website's server is designed to handle up to 550 vists per hour.


What is the probability of the website server being overwhelmed?

In [6]:
from scipy.stats import poisson

mu = 500 #average visits per hour
poisson_dist = poisson(mu)

result = 1 - poisson_dist.cdf(550) #more than550

print(f"Probability of the server being overwhelmed: {result:.2f}")

Probability of the server being overwhelmed: 0.01


What is the probability of being overwhelmed at some point during a day? (consider 24hours)

In [7]:
from scipy.stats import poisson, binom

mu = 500
poisson_dist = poisson(mu)
p_hour = 1 - poisson_dist.cdf(550) # probability in one hour of being overwhelmed, as calculated previously

In [8]:
n = 24 #24hours
binom_dist = binom(n, p_hour)
result = 1 - binom_dist.cdf(0)

print(f"Probability that one given hour is overwhelmed: {p_hour:.2f}")
print(f"Probability of being overwhelmed at some point during the day: {result:.2f}")

Probability that one given hour is overwhelmed: 0.01
Probability of being overwhelmed at some point during the day: 0.27


**Challenge 4**

#### Ironhack Helpdesk

Consider a scenario related to the time between arrivals of customers at a service desk.

On average, a customers arrives every 10minutes.

What is the probability that the next customer will arrive within the next 5 minutes?

In [9]:
#when time -> exponencial distribution

from scipy.stats import expon

expon_dist = expon(scale=10)  #every 10min one customer, so scale=10

result = expon_dist.cdf(5) #wihin 5min

print(f"Probability that the next customer arrives within 5 minutes: {result:.2f}")

Probability that the next customer arrives within 5 minutes: 0.39


If there is no customer for 15minutes, employees can that a 5minutes break.

What is the probability an employee taking a break?

In [14]:
from scipy.stats import expon

expon_dist = expon(scale=10)  #every 10min one customer, so scale=10

result = 1 - expon_dist.cdf(15)  #bc if no customer arrives for 15min, so t > 15 = right side

print(f"Probability of the employee taking a break: {result:.2f}")

Probability of the employee taking a break: 0.22


**Challenge 5**

The weights of a certain species of birds follow a normal distribution with a mean weight of 150 grams and a standard deviation of 10 grams. 

- If we randomly select a bird, what is the probability that its weight is between 140 and 160 grams?

In [15]:
# mean = 150g, std = 10g
# (valor − média) / desvio-padrão 
## (140 - 150) / 10 = -1
## (160 - 150) / 10 = 1
# 140 and 160 are exactly -1 and +1 standard deviations from the mean

In [16]:
from scipy.stats import norm

mean = 150
std = 10

z1 = (140 - mean) / std  
z2 = (160 - mean) / std   

result = norm.cdf(z2) - norm.cdf(z1)

print(f"Probability that the bird weighs between 140g and 160g: {result:.2f}")

Probability that the bird weighs between 140g and 160g: 0.68


**Challenge 6**

If the lifetime (in hours) of a certain electronic component follows an exponential distribution with a mean lifetime of 50 hours, what is the probability that the component fails within the first 30 hours?

In [18]:
from scipy.stats import expon

expon_dist = expon(scale=50)  

result = expon_dist.cdf(30) 

print(f"Probability that the electroic component fails within the first 30h is: {result:.2f}")

Probability that the electroic component fails within the first 30h is: 0.45
